### PR TITLE
feat: CDC integration

### PR DESCRIPTION
**Pull request by pratikkumar90**
Changes to support CDC in Spartacus. 

Create a new library project which contains, 
1. GigyaJSService - Which initializes on app startup and it handles user login and profile update functionality. 
2. GigyaRaasComponent - Component responsible for rendering the CDC Screen Sets 
3. Auth Module - New Auth module to call the new OAuth flow to fetch OAuth token-based on the data retrieved from the onLogin event from CDC
4. Enhancement to the sample storefrontapp to test the CDC functionality.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -87,7 +87,9 @@
     "build:cds:ssr": "cross-env SPARTACUS_CDS=true ng run storefrontapp:server:production",
     "prerender:cds": "cross-env SPARTACUS_CDS=true ng run storefrontapp:prerender",
     "release:cds:with-changelog": "cd projects/cds && release-it && cd ../..",
-    "start:b2b:ci:2005": "cross-env SPARTACUS_BASE_URL=https://dev-com-7.accdemo.b2c.ydev.hybris.com:9002 SPARTACUS_API_PREFIX=/occ/v2/ SPARTACUS_B2B=true ng serve --prod"
+    "release-notes:cds": "ts-node ./scripts/changelog.ts --verbose --lib cds",
```
Comment (marlass) [inline package.json]: I think this should be removed

File: package.json [hunk 2]
```diff
@@ -105,7 +107,7 @@
     "@compodoc/compodoc": "^1.1.10",
     "@ng-bootstrap/ng-bootstrap": "^6.0.0",
     "@ng-select/ng-select": "^4.0.0",
-    "@ngrx/effects": "~9.0.0",
+    "@ngrx/effects": "^9.0.0",
```
Comment (marlass) [inline package.json]: Why we change type of effect allowed versions?

File: projects/gigya/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Gigya
+
+Gigya library provides capabilities to integrate with SAP Customer Data Cloud from Gigya for enhanced identity management capabilities.
+
+`import { GigyaModule } from '@spartacus/gigya';`
```
Comment (marlass) [inline projects/gigya/README.md]: You can remove that line, as it will be described in details in the docs, how to enable it and configure.

**Review by marlass — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (marlass) [reply package.json]: That might be the reason behind the yarn.lock issue on Travis

File: projects/gigya/package.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "storefront",
+    "customer data cloud",
+    "gigya",
+    "cdc"
+  ],
+  "license": "Apache-2.0",
+  "publishConfig": {
+    "access": "public"
+  },
+  "repository": "https://github.com/SAP/spartacus",
```
Comment (marlass) [inline projects/gigya/package.json]: Can you point here to specific library directory? Similar thing like we do for other libraries?

File: projects/gigya/src/auth/store/actions/gigya-user-token.action.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+export class LoadUserTokenFail implements Action {
+  readonly type = LOAD_USER_TOKEN_FAIL;
+  constructor(public payload: ErrorModel | HttpErrorModel | any) {}
+}
+
+export class LoadGigyaUserToken implements Action {
+  readonly type = LOAD_GIGYA_USER_TOKEN;
+  constructor(
+    public payload: {
```
Comment (marlass) [inline projects/gigya/src/auth/store/actions/gigya-user-token.action.ts]: Could you extract it into interface and then use in LoadUserTokenSuccessPayload interface?

File: projects/gigya/src/auth/store/actions/gigya-user-token.action.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { Action } from '@ngrx/store';
+import { UserToken, ErrorModel, HttpErrorModel } from '@spartacus/core';
+
+export const LOAD_GIGYA_USER_TOKEN = '[Auth] Load Gigya User Token';
+export const LOAD_USER_TOKEN_FAIL = '[Auth] Load User Token Fail';
+export const LOAD_USER_TOKEN_SUCCESS = '[Auth] Load User Token Success';
+
+export interface LoadUserTokenSuccessPayload {
```
Comment (marlass) [inline projects/gigya/src/auth/store/actions/gigya-user-token.action.ts]: It needs to be exported?

File: projects/gigya/src/auth/store/effects/gigya-user-token.effect.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+              initActionPayload: payload,
+            });
+          }),
+          catchError((error) => {
+            this.globalMessageService.add(
+              { key: 'httpHandlers.badGateway' },
+              GlobalMessageType.MSG_TYPE_ERROR
+            );
+            return of(
+              new GigyaAuthActions.LoadUserTokenFail(
```
Comment (marlass) [inline projects/gigya/src/auth/store/effects/gigya-user-token.effect.ts]: Could you also pass initActionPayload to this action?

File: projects/gigya/src/auth/store/effects/index.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import { GigyaUserTokenEffects } from './gigya-user-token.effect';
+
+export const effects: any[] = [GigyaUserTokenEffects];
```
Comment (marlass) [inline projects/gigya/src/auth/store/effects/index.ts]: Remove this extra variable `effects`. Directly list `GigyaUserTokenEffects` in module

File: projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  initialize(): void {
+    this.loadGigyaJavascript();
+  }
+
+  /**
+   * Method which loads the CDC Script
+   */
+  loadGigyaJavascript(): void {
+    this.subscription.add(
+      this.baseSiteService.getActive().subscribe((baseSite) => {
```
Comment (marlass) [inline projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts]: Could you combine these two subscriptions with `combineLatest`?

File: projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        let javascriptUrl = this.getJavascriptUrlForCurrentSite(baseSite);
+        this.languageService.getActive().subscribe((language) => {
+          javascriptUrl = javascriptUrl + '&lang=' + language;
+          this.externalJsFileLoader.load(javascriptUrl, undefined, () => {
+            this.loaded$.next(true);
+            this.registerEventListeners();
```
Comment (marlass) [inline projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts]: You can pass here the basesite to registerEventListeners. Then you can remove `getCurrentBaseSite` method.

File: projects/gigya/src/auth/auth.module.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+import { HttpClientModule } from '@angular/common/http';
+import { NgModule } from '@angular/core';
+import { GigyaUserAuthenticationTokenService } from './services/user-authentication/gigya-user-authentication-token.service';
+import { GigyaAuthStoreModule } from './store/gigya-auth-store.module';
+
+@NgModule({
+  imports: [CommonModule, HttpClientModule, GigyaAuthStoreModule],
+  providers: [GigyaUserAuthenticationTokenService],
+})
+export class AuthModule {}
```
Comment (marlass) [inline projects/gigya/src/auth/auth.module.ts]: Could we rename it to `GigyaAuthModule`?

File: projects/gigya/src/gigya.module.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+export function gigyaJSFactory(gigyaJsService: GigyaJsService) {
+  const func = () => gigyaJsService.initialize();
+  return func;
+}
+
+@NgModule({
+  imports: [
+    GigyaRaasModule,
+    AuthModule,
+    EffectsModule.forRoot([GigyaUserTokenEffects]),
```
Comment (marlass) [inline projects/gigya/src/gigya.module.ts]: Effects are registered both here and in StoreModule. Is it intended?

File: package.json [hunk 13]
```diff
@@ -87,6 +87,8 @@
     "build:cds:ssr": "cross-env SPARTACUS_CDS=true ng run storefrontapp:server:production",
     "prerender:cds": "cross-env SPARTACUS_CDS=true ng run storefrontapp:prerender",
     "release:cds:with-changelog": "cd projects/cds && release-it && cd ../..",
+    "start:b2b:ci:2005": "cross-env SPARTACUS_BASE_URL=https://dev-com-7.accdemo.b2c.ydev.hybris.com:9002 SPARTACUS_API_PREFIX=/occ/v2/ SPARTACUS_B2B=true ng serve --prod",
```
Comment (marlass) [inline package.json]: This script can be removed

File: projects/gigya/src/auth/store/reducers/index.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+export const reducerToken: InjectionToken<ActionReducerMap<
+  GigyaAuthState
+>> = new InjectionToken<ActionReducerMap<GigyaAuthState>>('AuthReducers');
+
+export const reducerProvider: Provider = {
+  provide: reducerToken,
+  useFactory: getReducers,
+};
+
+export function clearAuthState(
```
Comment (marlass) [inline projects/gigya/src/auth/store/reducers/index.ts]: Do we have to repeat all this from the core module? I thought we only needed new reducer?

File: tsconfig.json [hunk 15]
```diff
@@ -22,6 +22,8 @@
       "@spartacus/assets": ["projects/assets/src/public_api"],
       "@spartacus/cds": ["projects/cds/public_api"],
       "@spartacus/incubator": ["projects/incubator/public_api"],
+      "@spartacus/gigya": ["projects/gigya/public_api"],
+      "@spartacus/schematics": ["projects/schematics/src/public_api"],
```
Comment (marlass) [inline tsconfig.json]: Why we have this duplicate?

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/gigya/src/gigya.module.ts (hunk 12 above)
Comment (pratikkumar90) [reply projects/gigya/src/gigya.module.ts]: No registering here is not required as it is already registered in the auth store. I will remove this.

File: projects/gigya/src/auth/store/effects/gigya-user-token.effect.spec.ts [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+      actions$ = hot('-a', { a: action });
+      const expected = cold('-b', { b: completion });
+
+      expect(userTokenEffect.loadGigyaUserToken$).toBeObservable(expected);
+      expect(testToken.expiration_time).toBeDefined();
+      expect(testToken.userId).toEqual(OCC_USER_ID_CURRENT);
+    });
+
+    it('should serialize string error before returning it', () => {
```
Comment (marlass) [inline projects/gigya/src/auth/store/effects/gigya-user-token.effect.spec.ts]: Could you extract all makeErrorSerializable tests into separate `describe` block?

File: projects/gigya/public_api.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Public API Surface for CDC
+ */
+export * from './src/gigya.module';
+export * from './src/cms-components/gigya-raas/index';
+export * from './src/auth/facade/gigya-auth.service';
+export * from './src/auth/store/actions/index';
```
Comment (marlass) [inline projects/gigya/public_api.ts]: GigyaJSService should also be exported in the public API. Additionally we should move this service into `facade` directory.

File: package.json [hunk 18]
```diff
@@ -87,11 +87,12 @@
...
+    "build:core:lib:gigya": "yarn build:core:lib && ng build gigya",
```
Comment (marlass) [inline package.json]: This script can be removed. Instead added it to general `build:core:lib`

File: projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+import { GigyaAuthService } from '../../auth/facade/gigya-auth.service';
+import { GigyaConfig } from '../../config/gigya-config';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class GigyaJsService implements OnDestroy {
+  private loaded$ = new ReplaySubject<boolean>(1);
+  private errorLoading$ = new ReplaySubject<boolean>(1);
+  subscription: Subscription = new Subscription();
```
Comment (marlass) [inline projects/gigya/src/cms-components/gigya-js/gigya-js.service.ts]: Let's make all these properties protected instead of private.

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/gigya/public_api.ts (hunk 17 above)
Comment (marlass) [reply projects/gigya/public_api.ts]: I meant to move the file under this path: `/src/auth/facade/gigya-js.service` as it is no longer related to CMS components.

**Review by marlass — APPROVED** [review]

**Pull request merged**